### PR TITLE
Use task sources for device events and canvas texture expiry

### LINKED ISSUE #3295: Canvas texture expiry timing is unpredictable
EDIT: Updated title to reflect both: issues with invisible canvases, and general unreliability of timing in the event loop.

---

Right now it's not clear whether [Update the rendering of the WebGPU canvas](https://gpuweb.github.io/gpuweb/#abstract-opdef-update-the-rendering-of-the-webgpu-canvas) gets called for canvases not in the DOM. This affects the advancement of the frame, i.e. whether the next call to `getCurrentTexture()` will return the same texture or a new one.

WebGL (with `preserveDrawingBuffer: false`) has an equivalent question, which is whether the drawing buffer gets replaced or not. This test case tries to test that: https://codepen.io/kainino0x/pen/mdxKKyM
Unfortunately, the results are kind of wonky.
- Firefox and Chrome seem to advance the frame only if the canvas is actually visible (inside the viewport) OR it's had transferControlToOffscreen called on it. Neither of them advances a `new OffscreenCanvas`.
- Safari seems to advances the frame as long as it's in the DOM. (no OffscreenCanvas support)

---

Tasks (see below)

- [ ] Figure out behavior for canvases
- [ ] Consider if behavior for external textures should match closer to behavior for canvases

### PR DESCRIPTION
**Pull request by kainino0x**
- Better define device loss, define task source for content timeline steps
- Define automatic expiry of canvas textures using a task source

Fixes #3295

### CODE REVIEW

File: spec/index.bs [hunk 1]
```diff
@@ -1970,6 +1981,55 @@ For more information on issuing CORS requests for image and video elements, cons
...
+    Note: Since task source priority is not normatively defined, there is a range of points at which
+    tasks from this task queue can execute. In practice, this means these tasks can be implemented
+    with fixed steps inside the [[HTML#event-loop-processing-model|HTML processing model]].
+    All of the following options are valid:
+
+    - Immediately after running any task.
+    - Immediately after running animation frame callbacks.
+    - Immediately before rendering.
+        (In the case of canvas texture expiry, this is already done in
+        [$update the rendering of the WebGPU canvas$], so the task is allowed to be omitted entirely.)
```
Comment (toji) [inline spec/index.bs:2030]: Recognizing that this is non-normative, I still feel like this phrasing is slightly risky. One of the reasons we set off on this whole "When does it expire" journey is that we found cases in Chrome where the rendering wasn't being performed (tab was in the background, etc) and so the texture never expired. I'd hate to have other implementations look at this and say "Ah, we'll just expire on composite and skip the whole scheduling thing" and be bitten by the same bug.

I'm not sure exactly how to phrase it, but I think I'd want this to state more clearly something along the lines of "If the texture is expired in [$update the rendering of the WebGPU canvas$] then a separate expiry task is not necessary, but a task should be scheduled if the canvas rendering is not updated."?

File: spec/index.bs [hunk 2]
```diff
@@ -12308,23 +12379,38 @@ interface GPUCanvasContext {
...
     1. If |context|.{{GPUCanvasContext/[[currentTexture]]}} is not `null`:
         1. Call |context|.{{GPUCanvasContext/[[currentTexture]]}}.{{GPUTexture/destroy()}}
             (without destroying |context|.{{GPUCanvasContext/[[drawingBuffer]]}})
             to terminate write access to the image.
         1. Set |context|.{{GPUCanvasContext/[[currentTexture]]}} to `null`.
 
+    Note:
+    This will be called multiple times for the same texture, in some order, from both
+    {{GPUCanvasContext/getCurrentTexture()}} and [$update the rendering of the WebGPU canvas$].
+    The second call has no effect.
```
Comment (toji) [inline spec/index.bs:12398]: Actually, re: my comment above, I wonder if this is enough? Just acknowledging that we might enter this algorithm multiple times for a single texture and that's OK? I'd feel better about that than indicating that the expiration step can be skipped.

**Review by toji — APPROVED** [review]
LGTM, with a minor concern about the non-normative text.

File: spec/index.bs (hunk 1 above)
Comment (kainino0x) [reply spec/index.bs:2030]: It's intended that that would be an allowed implementation. Wrote a warning into the note while trying not to say it's disallowed.

File: spec/index.bs [hunk 3]
```diff
@@ -1970,6 +1981,60 @@ For more information on issuing CORS requests for image and video elements, cons
...
+    - Immediately after running animation frame callbacks.
+    - Immediately before rendering.
+        Note this choice results in less predictable expiry, especially for non-visible canvases
+        (e.g. scrolled off the screen or in a background tab), potentially allowing invalid programs
+        to appear correct or changing the behavior of websites which (incorrectly) rely on expiry
+        timing.
+
+        In the case of canvas texture expiry, expiry is also done in [$update the rendering of
+        the WebGPU canvas$], so it's equivalent for an implementation to elide the task entirely.
+n</div>
```
Comment (shrekshao) [inline spec/index.bs:2036]: Typo 'n'

**Pull request merged**